Initial docker-compose setup (WIP)

garry-jeromson opened the pull request:
This is far from complete, and is meant to serve as a starting point for discussion. 

A Dockerized development setup has the advantage that all the developer needs to have installed is `make` and Docker - all the Python stuff is containerised, making it more accessible (and hopefully, because of that, more "developable"). This, however, is a departure from the current project setup based on tox. 

Tox and Docker are two difference approaches to the same problem (testing with multiple Python versions) - it doesn't really make sense to run tox commands within a Docker container, and it may be tricky to support both development approaches simultaneously.

This commit forms the basis of my suggested initial approach:

- Development requirements should be specified in a requirements-dev.txt file, to allow easy installation outside of tox
- Runtime requirements should ideally be specified in requirements.txt, to be available outside of tox (assuming tox and setup.py can then reference it)
- docker-compose is used to manage Python versions, each with a separate file. This means more files, but reduced Makefile complexity
- The Dockerfile for the development image takes the Python version as a build argument, to allow easily adding new versions.

brunns commented on setup.py:
```diff
@@ -18,6 +18,9 @@
     logger.warning("README file not found or unreadable.")
     long_description = "See https://github.com/brunns/mbtest/"
 
+with open('requirements.txt') as f:
```
Is there some way we can do without this? This is how I used to do it, but there is at least one downstream service which needs `setup.py` to be self contained, IIRC.

garry-jeromson replied:
Might be that we just have to live with having two copies of the requirements list - not really a biggie.

garry-jeromson replied:
Actually, if `tox.ini`, `setup.py` and `requirements.txt` can live independently of each other, it may simplify supporting both non-Dockerised and Dockerised workflows - the Docker containers would rely on the `requirements.txt` files, and tox would keep on doing its thing. Do you see any potential issues there?

brunns replied:
I'd like something in the build ensuring the two are in sync. If we have that, I can live with the duplication.

brunns replied:
Or, better still, can we build `requirements.txt` from `setup.py`, make it a generated artifact?

garry-jeromson replied:
One potential option: https://github.com/jazzband/pip-tools#example-usage-for-pip-compile

brunns replied:
Looks worth a try!

brunns replied:
[Travis](https://travis-ci.org/github/brunns/mbtest) is unhappy. I'm not sure why that's not showing up in the PR - I'll look into that.

garry-jeromson commented:
I'd say we're cooking with gas on this now - all that's left to do is to bring the Docker Makefile to feature parity with the tox one, and update the docs.

brunns commented:
Can you add a travis target, so we can see if anything breaks the docker build?

brunns commented:
Is the `Makefile` rename necessary? I'd like to continue to be able to just run `make precommit` before committing.

garry-jeromson commented:
> Is the `Makefile` rename necessary? I'd like to continue to be able to just run `make precommit` before committing.

Have changed it back and added an easy way to switch between the two from the command line.

> Can you add a travis target, so we can see if anything breaks the docker build?

I'm a Travis noob - not sure if what I added is in the right place.

brunns commented on Makefile:
```diff
@@ -88,6 +88,11 @@ repl: ## Python REPL
 	tox -e py38 -- pip list --outdated
 
+.PHONY: use-docker
+use-docker: ## Use Docker-based local development workflow
```
This leaves my `git status` _dirty_. I wonder if that's avoidable?

garry-jeromson replied:
An alternative would be to have both tox and Docker targets in the same Makefile, and switching which targets are executed based on a variable. That would still leave the variable as a potential dirty status candidate though, and the Makefile would be really long.

brunns commented:
How are we going with this? BTW, I came across https://github.com/kudulab/dojo which might be useful...

garry-jeromson commented:
Coming back to this after a loooong time with fresh eyes; I wonder if it's actually going to be useful to anyone. The tox-based workflow isn't really a hurdle at all if you know what you're doing, and the Docker stuff adds complexity - if it's not gonna be used by other folks, perhaps we should just can it.

brunns commented:
Perhaps. I won't be using it myself, put perhaps it'll be useful to some folks?
